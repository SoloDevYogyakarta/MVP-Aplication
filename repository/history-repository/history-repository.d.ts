import { UserInstance } from '../../database/entities/authenticate/user-entity/user-entity';
export declare class HistoryRepository {
    private readonly logger;
    findAll(): Promise<UserInstance[]>;
    findOne(id: number): Promise<UserInstance>;
    visit(id: number): Promise<number>;
}
