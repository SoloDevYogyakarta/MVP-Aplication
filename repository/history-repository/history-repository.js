"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var HistoryRepository_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.HistoryRepository = void 0;
const common_1 = require("@nestjs/common");
const entity_1 = require("../../database/entities/entity");
const user_associate_1 = require("../../database/associates/user-associate/user-associate");
const filter_1 = require("../../utils/filter/filter");
let HistoryRepository = HistoryRepository_1 = class HistoryRepository {
    constructor() {
        this.logger = new common_1.Logger(HistoryRepository_1.name);
    }
    async findAll(query, type) {
        let where = {};
        this.logger.log(HistoryRepository_1.name);
        where = (0, filter_1.filter)(query, type);
        let result = await user_associate_1.userAssociate.findAll({
            where: where,
            attributes: user_associate_1.userAttribute,
            include: user_associate_1.userHistoryInclude,
            order: [['id', 'ASC']],
        });
        for (const values of result) {
            const visit = await this.visit(values.id);
            const total = await this.total(values.id);
            const totalOrder = await this.totalOrder(values.id);
            result = JSON.parse(JSON.stringify(result)).map((item) => {
                if (item.id === values.id) {
                    item['visit'] = Number(visit);
                    item['total'] = total;
                }
                if (totalOrder.length) {
                    item['order'] = item.order.map((ord) => {
                        const total_order = totalOrder.find((i) => i.id === ord.id);
                        if (total_order) {
                            if (item.id === ord.user_id) {
                                ord['total'] = total_order?.total
                                    ? Number(total_order.total)
                                    : 0;
                            }
                        }
                        return ord;
                    });
                }
                return item;
            });
        }
        return result;
    }
    async findOne(id) {
        this.logger.log(HistoryRepository_1.name);
        let result = await user_associate_1.userAssociate.findOne({
            where: { id },
            attributes: user_associate_1.userAttribute,
            include: user_associate_1.userHistoryInclude,
        });
        result = JSON.parse(JSON.stringify(result));
        for (const values of result
            .order) {
            const total = await this.totalOrder(values.id);
            result['order'] = result.order.map((item) => {
                if (item.id === values.id) {
                    item['total'] = total[0].total;
                }
                return item;
            });
        }
        result['visit'] = Number(await this.visit(id));
        result['total'] = await this.total(id);
        return result;
    }
    async visit(id) {
        const query = await entity_1.sequelize.query(`SELECT a.user_id, COUNT(a.user_id) as visit FROM "PRODUCTS"."ORDER" AS a WHERE a.user_id = '${id}' GROUP BY a.user_id `);
        const visit = query[0].find((item) => Number(item.user_id) === Number(id));
        return visit?.visit ? visit.visit : 0;
    }
    async totalOrder(id) {
        const query = await entity_1.sequelize.query(`
    SELECT a.id,a.user_id,SUM(CAST(b.price as int)) as total FROM "PRODUCTS"."ORDER" AS a
    LEFT JOIN "SERVICES"."HISTORY" AS b ON b.order_id = a.id WHERE a.user_id = ${id}
    GROUP BY a.id;
    `);
        return query[0];
    }
    async total(id) {
        const query = await entity_1.sequelize.query(`
      SELECT a.user_id,SUM(CAST(b.price as int)) as total FROM "PRODUCTS"."ORDER" AS a
      LEFT JOIN "SERVICES"."HISTORY" AS b ON b.order_id = a.id WHERE a.user_id = ${id}
      GROUP BY a.id
    `);
        const result = query[0]
            .map((item) => (item.total === null ? 0 : item.total))
            ?.reduce((arr, curr) => {
            return Number(arr) + Number(curr);
        }, 0);
        return result ? Number(result) : 0;
    }
};
exports.HistoryRepository = HistoryRepository;
exports.HistoryRepository = HistoryRepository = HistoryRepository_1 = __decorate([
    (0, common_1.Injectable)()
], HistoryRepository);
//# sourceMappingURL=history-repository.js.map