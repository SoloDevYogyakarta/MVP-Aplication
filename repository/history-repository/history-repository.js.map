{"version":3,"file":"history-repository.js","sourceRoot":"","sources":["../../../src/repository/history-repository/history-repository.ts"],"names":[],"mappings":";;;;;;;;;;AAAA,2CAAoD;AACpD,2DAA2D;AAC3D,4FAIiE;AAI1D,IAAM,iBAAiB,yBAAvB,MAAM,iBAAiB;IAAvB;QACY,WAAM,GAAG,IAAI,eAAM,CAAC,mBAAiB,CAAC,IAAI,CAAC,CAAC;IAyC/D,CAAC;IAvCC,KAAK,CAAC,OAAO;QACX,IAAI,CAAC,MAAM,CAAC,GAAG,CAAC,mBAAiB,CAAC,IAAI,CAAC,CAAC;QACxC,IAAI,MAAM,GAAG,MAAM,8BAAa,CAAC,OAAO,CAAC;YACvC,UAAU,EAAE,8BAAa;YACzB,OAAO,EAAE,mCAAkB;SAC5B,CAAC,CAAC;QACH,KAAK,MAAM,MAAM,IAAI,MAAM,EAAE;YAC3B,MAAM,KAAK,GAAG,MAAM,IAAI,CAAC,KAAK,CAAC,MAAM,CAAC,EAAE,CAAC,CAAC;YAC1C,MAAM,GAAG,IAAI,CAAC,KAAK,CAAC,IAAI,CAAC,SAAS,CAAC,MAAM,CAAC,CAAC,CAAC,GAAG,CAAC,CAAC,IAAkB,EAAE,EAAE;gBACrE,IAAI,IAAI,CAAC,EAAE,KAAK,MAAM,CAAC,EAAE,EAAE;oBACzB,IAAI,CAAC,OAAO,CAAC,GAAG,KAAK,CAAC;iBACvB;gBACD,OAAO,IAAI,CAAC;YACd,CAAC,CAAC,CAAC;SACJ;QACD,OAAO,MAAM,CAAC;IAChB,CAAC;IAED,KAAK,CAAC,OAAO,CAAC,EAAU;QACtB,IAAI,CAAC,MAAM,CAAC,GAAG,CAAC,mBAAiB,CAAC,IAAI,CAAC,CAAC;QACxC,IAAI,MAAM,GAAG,MAAM,8BAAa,CAAC,OAAO,CAAC;YACvC,KAAK,EAAE,EAAE,EAAE,EAAE;YACb,UAAU,EAAE,8BAAa;YACzB,OAAO,EAAE,mCAAkB;SAC5B,CAAC,CAAC;QACH,MAAM,GAAG,IAAI,CAAC,KAAK,CAAC,IAAI,CAAC,SAAS,CAAC,MAAM,CAAC,CAAC,CAAC;QAC5C,MAAM,CAAC,OAAO,CAAC,GAAG,MAAM,IAAI,CAAC,KAAK,CAAC,EAAE,CAAC,CAAC;QACvC,OAAO,MAAM,CAAC;IAChB,CAAC;IAED,KAAK,CAAC,KAAK,CAAC,EAAU;QACpB,MAAM,KAAK,GAAG,MAAM,kBAAS,CAAC,KAAK,CACjC,4EAA4E,EAAE,GAAG,CAClF,CAAC;QACF,MAAM,KAAK,GAAG,KAAK,CAAC,CAAC,CAAC,CAAC,IAAI,CACzB,CAAC,IAAwC,EAAE,EAAE,CAAC,IAAI,CAAC,OAAO,KAAK,EAAE,CAC5B,CAAC;QACxC,OAAO,KAAK,EAAE,KAAK,CAAC,CAAC,CAAC,KAAK,CAAC,KAAK,CAAC,CAAC,CAAC,CAAC,CAAC;IACxC,CAAC;CACF,CAAA;AA1CY,8CAAiB;4BAAjB,iBAAiB;IAD7B,IAAA,mBAAU,GAAE;GACA,iBAAiB,CA0C7B"}