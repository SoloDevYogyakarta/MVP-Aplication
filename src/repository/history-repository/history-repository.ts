import { Injectable, Logger } from '@nestjs/common';
import {
  userAssociate,
  userAttribute,
  userHistoryInclude,
} from '../../database/associates/user-associate/user-associate';
import { UserInstance } from '../../database/entities/authenticate/user-entity/user-entity';

@Injectable()
export class HistoryRepository {
  private readonly logger = new Logger(HistoryRepository.name);

  async findAll(): Promise<UserInstance[]> {
    this.logger.log(HistoryRepository.name);
    return await userAssociate.findAll({
      attributes: userAttribute,
      include: userHistoryInclude,
    });
  }

  async findOne(public_id: string): Promise<UserInstance> {
    this.logger.log(HistoryRepository.name);
    return await userAssociate.findOne({
      where: { public_id },
      attributes: userAttribute,
      include: userHistoryInclude,
    });
  }
}
