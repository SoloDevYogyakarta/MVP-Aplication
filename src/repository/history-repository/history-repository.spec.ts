import { HistoryRepository } from './history-repository';

describe('HistoryRepository', () => {
  let repository: HistoryRepository;

  beforeEach(() => {
    repository = new HistoryRepository();
  });

  it('should to be defined', () => expect(repository).toBeDefined());

  it('render correctly', () => expect(repository).toMatchSnapshot());
});
