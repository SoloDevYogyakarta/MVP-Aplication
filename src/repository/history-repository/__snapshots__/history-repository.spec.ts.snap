// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HistoryRepository render correctly 1`] = `
HistoryRepository {
  "logger": Logger {
    "context": "HistoryRepository",
    "options": {},
  },
}
`;
